src/Node_verify.cpp

```diff
@@ -354,6 +354,212 @@ void Node::verify_vdf(std::shared_ptr<const ProofOfTime> proof, const uint32_t c
+	const uint32_t batch_size = (num_point_cpu + async_num - 1) / async_num;
+	const uint32_t batch_full = (num_point_cpu % async_num) ? (num_point_cpu % async_num) : async_num;
+	// TODO: check/fix logic if num_point_cpu less than async_num
+
+	// comment: store start/stop timestamp for all cpu threads, 1x opencl thread
+	std::vector<uint64_t> time_cpu_start;
+	std::vector<uint64_t> time_cpu_stop;
+	uint64_t time_ocl;
+	time_cpu_start.resize(async_num);
+	time_cpu_stop.resize(async_num);
+
+	// comment: hash all points, load balanced on cpu/opencl
+#pragma omp parallel for
+	for(int chunk = 0; chunk < (int)async_num + 1; ++chunk)
+	{
+		if(chunk < (int)async_num) {
+			// comment: cpu, async_num threads
+			time_cpu_start[chunk] = vnx::get_time_micros();
+			const uint32_t batch_real = (chunk < (int)batch_full) ? batch_size : batch_size - 1;
+			const uint32_t batch_startp = (chunk < (int)batch_full) ? (chunk * batch_size) : (batch_full * batch_size) + ((chunk - batch_full) * (batch_size - 1));
+
+			if(have_sha_ni) {
+				// comment: cpu, sha-ni
+				for(uint32_t j = 0; j < batch_real; ++j)
+				{
+					const uint32_t i = batch_startp + j;
+					recursive_sha256_ni(point[i].data(),num_iters[i]);
+				}
+			} else {
+				// comment: cpu, avx2
+				constexpr uint32_t avx_batch_size = 8;
+				const uint32_t avx_num_chunks = (batch_real + avx_batch_size - 1) / avx_batch_size;
+				const uint32_t avx_batch_lastsize = (batch_real % avx_batch_size) ? batch_real % avx_batch_size : avx_batch_size;
+				// TODO: check/fix logic if batch_real less than avx_batch_size
+
+				for(uint32_t avx_chunk = 0; avx_chunk < avx_num_chunks; ++avx_chunk)
+				{
+					const uint32_t avx_batch_real = (avx_chunk < avx_num_chunks - 1) ? avx_batch_size : avx_batch_lastsize;
+					uint32_t max_iters = 0;
+					uint8_t hash[avx_batch_size][32];
+					uint8_t input[avx_batch_size][64];
+
+					for(uint32_t j = 0; j < avx_batch_real; ++j)
+					{
+						const uint32_t i = (batch_startp) + (avx_chunk * avx_batch_size) + j;
+						max_iters = std::max(max_iters, num_iters[i]);
+					}
+
+					for(uint32_t k = 0; k < max_iters; ++k)
+					{
+						for(uint32_t j = 0; j < avx_batch_real; ++j)
+						{
+							const uint32_t i = (batch_startp) + (avx_chunk * avx_batch_size) + j;
+							max_iters = std::max(max_iters, num_iters[i]);
+							::memcpy(input[j], point[i].data(), 32);
+						}
+						// comment: if testing avx2, call direct sha256_avx2_64_x8(), or else sha-ni
+						sha256_64_x8(hash[0],input[0],32);
+
+						for(uint32_t j = 0; j < avx_batch_real; ++j)
+						{
+							const uint32_t i = (batch_startp) + (avx_chunk * avx_batch_size) + j;
+							if(k < num_iters[i]) {
+								::memcpy(point[i].data(), hash[j], 32);
+							}
+						}
+					}
+				}
+			}
+
+			time_cpu_stop[chunk] = vnx::get_time_micros();
+		} else {
+			// comment: opencl, 1x thread
+			time_ocl = vnx::get_time_micros();
+			auto engine = opencl_vdf[0];
```

Comment:
The reason there's 3 OpenCL instances is that on big GPUs it allows to compute all three streams in parallel, since a single stream typically does not fully utilize a big GPU (a single stream is only ~1140 cores).

Reply:
But I suppose the use case for both CPU and GPU is only when you got a weak GPU anyways, so using just one instance here is probably fine.

Reply:
Thx for info (learing). Will rework to 3x OpenCL instances, depending on new variant of load balance.